tensorflow.go

```diff
@@ -0,0 +1,172 @@
+	tf "github.com/tensorflow/tensorflow/tensorflow/go"
+)
+
+type Session struct {
+	session tf.TF_Session
+}
+
+func NewSession() (*Session, error) {
+	status := tf.TF_NewStatus()
+	return &Session{
+		session: tf.TF_NewSession(
+			tf.TF_NewSessionOptions(),
+			status,
+		),
+	}, statusToError(status)
+}
+
+type Tensor struct {
+	tensor tf.TF_Tensor
+	buf    []byte
+}
+
+type TensorShape [][]int64
+
+var (
+	TensorShapeScalar = TensorShape{{1}}
+)
+
+func NewTensor(dataType tf.DataType, shape TensorShape, data interface{}) (*Tensor, error) {
+	// TODO(tmc): ensure data is a slice
+	v := reflect.ValueOf(data)
+	if v.Kind() != reflect.Slice {
+		return nil, fmt.Errorf("tensorflow: 'data' argument must be a slice")
+	}
+	dataSize := v.Len() * int(v.Type().Elem().Size())
+	dataPtr := v.Pointer()
+	return newTensor(dataType, shape, dataPtr, dataSize)
+}
+
+func newTensor(dataType tf.DataType, shape TensorShape, data uintptr, size int) (*Tensor, error) {
+	t := &Tensor{
+		tensor: tf.TF_NewTensor_wrapper(tf.TF_DataType(dataType), &(shape[0][0]), len(shape), data, int64(size)),
+	}
+
+	return t, nil
+}
+
+func encodeStrings(in []string) []byte {
+	size := 0
+	for _, s := range in {
+		size += 8
+		size += len(s)
+		size += len(proto.EncodeVarint(uint64(len(s))))
+	}
+
+	out := make([]byte, size)
+
+	dataPos := 8 * len(in)
+	data := out[dataPos:]
+	offset := 0
+	for i, s := range in {
+		inBytes := []byte(s)
+		binary.LittleEndian.PutUint64(out[i*8:], uint64(offset))
+		inLen := proto.EncodeVarint(uint64(len(s)))
+		offset += copy(data[offset:], inLen)
+		offset += copy(data[offset:], inBytes)
+	}
+	return out
+}
+
+func Constant(value interface{}) (*Tensor, error) {
+	switch v := value.(type) {
+	case string:
+		//		defer C.free(unsafe.Pointer(str))
+		buf := encodeStrings([]string{v})
+		t, err := newTensor(tf.DataType_DT_STRING, TensorShapeScalar, uintptr(unsafe.Pointer(&(buf[0]))), len(buf))
+		if err != nil {
+			return nil, err
+		}
+		t.buf = buf
+		return t, nil
+	default:
+		return nil, fmt.Errorf("tensorflow: unsupported type %T", value)
+	}
+}
+
+func statusToError(status tf.TF_Status) error {
+	code := tf.TF_GetCode(status)
+	message := tf.TF_Message(status)
+
+	if code != 0 {
+		return fmt.Errorf("tensorflow: %d: %v", code, message)
+	}
+	return nil
+}
+
+func (s *Session) Run(inputs map[string]*Tensor, outputs []string, targets []string) ([]*Tensor, error) {
+	inputNames := tf.NewStringVector()
+	inputValues := tf.NewTensorVector()
+	if inputs != nil {
+		for k, v := range inputs {
+			inputValues.Add(v.tensor)
+			inputNames.Add(k)
+		}
+	}
+	outputNames := tf.NewStringVector()
+	for _, n := range outputs {
+		outputNames.Add(n)
+	}
+
+	targetNames := tf.NewStringVector()
+	for _, n := range targets {
+		targetNames.Add(n)
+	}
+
+	outputValues := tf.NewTensorVector()
+	status := tf.TF_NewStatus()
+
+	tf.TF_Run_wrapper(s.session, inputNames, inputValues, outputNames, outputValues, targetNames, status)
+
+	result := []*Tensor{}
```

Comment:
Because you know the size (outputValues.Size()) in advance, I'd be tempted to make() the result the right size and directly set result[i] in the below list, but this is relatively minor.